### PR TITLE
advanged-login

### PR DESCRIPTION
**Pull request by monleopa**

### CODE REVIEW

**Comment by monleopa** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -76,7 +76,7 @@ p {
   padding-top: 9px;
   font-weight: bold;
   &:hover {
-    color: white;
+    color: red;
```
Comment (huuphuc93) [inline app/assets/stylesheets/custom.scss]: use color variable

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -171,3 +171,17 @@ input {
...
+  span {
+    margin-left: 20px;
+    font-weight: normal;
+  }
+}
+
+#session_remember_me {
+  width: auto;
+  margin-left: 0;
+}
```
Comment (huuphuc93) [inline app/assets/stylesheets/custom.scss]: end of file

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -3,15 +3,37 @@ def log_in user
...
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by(id: user_id)
+    elsif user_id = cookies.signed[:user_id]
+      raise       # The tests still pass, so this branch is currently untested.
```
Comment (huuphuc93) [inline app/helpers/sessions_helper.rb]: what this?

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -3,15 +3,37 @@ def log_in user
...
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by(id: user_id)
```
Comment (huuphuc93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -3,15 +3,37 @@ def log_in user
...
+    elsif user_id = cookies.signed[:user_id]
+      raise       # The tests still pass, so this branch is currently untested.
+      user = User.find_by(id: user_id)
```
Comment (huuphuc93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -3,15 +3,37 @@ def log_in user
...
+      if user && user.authenticated?(cookies[:remember_token])
```
Comment (huuphuc93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -3,15 +3,37 @@ def log_in user
...
+    end
   end
 
   def logged_in?
     current_user.present?
   end
 
+  def forget user
+    user.forget
+    cookies.delete(:user_id)
```
Comment (huuphuc93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/models/user.rb [hunk 8]
```diff
@@ -13,4 +14,33 @@ def User.digest string
...
     BCrypt::Password.create(string, cost: cost)
   end
+
+  class << self
+    def digest string
+      cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+                                                    BCrypt::Engine.cost
+      BCrypt::Password.create(string, cost: cost)
+    end
+    def new_token
```
Comment (huuphuc93) [inline app/models/user.rb]: add and empty line between 2 methods

File: app/models/user.rb [hunk 9]
```diff
@@ -13,4 +14,33 @@ def User.digest string
...
+    update_attributes remember_digest: User.digest remember_token
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+
+  def forget
+    update_attributes remember_digest: nil)
```
Comment (huuphuc93) [inline app/models/user.rb]: what this )

File: app/views/sessions/new.html.erb [hunk 10]
```diff
@@ -8,6 +8,10 @@
       <%= f.email_field :email, class: "form-control" %>
       <%= f.label :password %>
       <%= f.password_field :password, class: "form-control" %>
+      <%= f.label :remember_me, class: "checkbox inline" do %>
+        <%= f.check_box :remember_me %>
+        <span>Remember me on this computer</span>
```
Comment (huuphuc93) [inline app/views/sessions/new.html.erb]: I18n

**Comment by huuphuc93** [comment]
commented
(tat) xóa lỗi này

**Comment by monleopa** [comment]
ready

File: app/helpers/sessions_helper.rb [hunk 11]
```diff
@@ -3,15 +3,37 @@ def log_in user
...
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by id: user_id
+    elsif user_id = cookies.signed[:user_id]
+      raise
+      user = User.find_by(id: user_id)
```
Comment (huuphuc93) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/models/user.rb [hunk 12]
```diff
@@ -13,4 +14,34 @@ def User.digest string
...
+                                                    BCrypt::Engine.cost
+      BCrypt::Password.create(string, cost: cost)
+    end
+    
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+  end
+
+  def User.new_token
```
Comment (huuphuc93) [inline app/models/user.rb]: duplicate method

**Comment by huuphuc93** [comment]
commented

**Comment by monleopa** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 13]
```diff
@@ -5,6 +5,7 @@ def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate(params[:session][:password])
       log_in user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (huuphuc93) [inline app/controllers/sessions_controller.rb]: use " instead of '

File: app/models/user.rb [hunk 14]
```diff
@@ -13,4 +14,30 @@ def User.digest string
...
+    end
+
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attributes remember_digest: User.digest(remember_token)
```
Comment (huuphuc93) [inline app/models/user.rb:32]: remove ()

**Comment by huuphuc93** [comment]
commented

File: app/models/user.rb (hunk 14 above)
Comment (monleopa) [reply app/models/user.rb:32]: sir, if i remove (), code is error

**Comment by monleopa** [comment]
ready

**Comment by huuphuc93** [comment]
commented 
please fix this comment
screenshot from 2018-08-01 08-21-31

File: app/controllers/sessions_controller.rb (hunk 13 above)
Comment (monleopa) [reply app/controllers/sessions_controller.rb]: i used " in the whole project except file .css . But when i push again the comment still exist

**Comment by monleopa** [comment]
ready

**Comment by huuphuc93** [comment]
merged
this (pr) can be (merged) now
CC: [To:2076564] Nguyen Huy Hung C (93)
CC: [To:1546746] Md. Mahmudur Rahman (Mahmud)
CC: [To:1818111] Nguyen Thi Huong (94)

**Pull request merged**